### ISSUE TITLE
Incorrect possible dereference of a null reference warning in case of 'x ?? throw' null check

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#32879]
**Version Used**: 

Microsoft Visual Studio Enterprise 2019 Preview
Version 16.0.0 Preview 2.0
VisualStudio.16.Preview/16.0.0-pre.2.0+28522.59
Microsoft .NET Framework
Version 4.7.03056

**Steps to Reproduce**:

Compile the following code
```cs
#nullable enable
class C
{
  void Test(string? s)
  {
    _ = s ?? throw new System.ArgumentNullException(nameof(s));
    s.ToString();
  }
}
```

**Expected Behavior**:
No warnings. `s` is explicitly checked for null and if it's null an exception is throw on the first line. It's impossible for `s` to hold a null value on the second line.

**Actual Behavior**:
Incorrect `Possible dereference of a null reference` warning for `s.ToString()`

**Comment by TessenR** [comment]
Similar case:
```cs
class C
{
  void M(string? s1)
  {
    var s2 = s1 ?? s1.ToString();
    s1.ToString(); // incorrect warning for possible dereference of a null reference
  }
}
```

**Comment by jcouv** [comment]
I think the second scenario behaves correctly. `s1` could still be `null`.

**Comment by TessenR** [comment]
@jcouv Could you please explain how it could still be null?
The expression `s1 ?? s1.ToString()` dereferences `s1` iff it's null therefore it's either not null (because left side of `??` returned not null) or it's dereferenced and thus cannot be null too (since `s1.ToString()` on the right side of `??` should update nullable state of `s1` after issuing a warning).

Consider the following equivalent examples:
```cs
#nullable enable
class C
{
  void M1(string? s1)
  {
    var s2 = s1 != null ? s1 : s1.ToString(); // s1 is either not null or dereferenced
    s1.ToString(); // no warnings here
  }

  void M2(string? s1)
  {
    string s2;
    if (s1 != null)
    {
      s2 = s1; // s1 is either not null
    }
    else
    {
      s1.ToString(); // or dereferenced
    }

    s1.ToString(); // no warnings issued here
  }
}
```

I've tried to keep the original example as simple as possible so it admittedly does not make much sense.

**Comment by jcouv** [comment]
You're correct. The first `s1.ToString()` should produce a warning and we assume `s1` isn't `null` in that branch. I'll add a test to verify. Thanks :-)

**Comment by jcouv** [comment]
Fixed for 16.1p1.
Note that LDM decided that `throw null` is a warning too (just like `throw maybeNullException`).
